src/translators/sql_translator.py

```diff
@@ -533,8 +534,13 @@ def _get_where_clause(self, entity_ids, from_date, to_date, fiware_sp=None,
 
         if fiware_sp:
             # Match prefix of fiware service path
-            clauses.append(
-                " " + FIWARE_SERVICEPATH + " ~* '" + fiware_sp + "($|/.*)'")
+            if fiware_sp == '/':
+                clauses.append(
+                    " " + FIWARE_SERVICEPATH + " ~* '($|/.*)'")
```

Comment:
previous query was not working correctly with "/" subpaths

Reply:
Not sure about this change. Let's start with an example.

```
create table t (path text);
insert into t values (''), ('a'), ('a/b'), ('a/b/c'), ('/'), ('/a'), ('/a/b'), ('/a/b/c');
select * from t where path ~* '($|/.*)';
```

this selects five paths: the empty path `''` and the four "absolute" (i.e. starting with a `/`) paths `/`, `/a`, `/a/b`, `/a/b/c`. Judging by the code, the empty path is sort of different from any other path in that it should only match an empty path in the DB? If so then an input path of `/` should not match the empty path...or should it? It looks to me the intention here is to match an input path `x` against the set of DB paths `FSP` as below

```
(1)   x == None or x == ''         ===>   { '' }
(2)   len(x) > 0 and x[0] != '/'   ===>   { p ∈ FSP | p starts with: '/' + x }
(3)   len(x) > 0 and x[0] == '/'   ===>   { p ∈ FSP | p starts with: x } 
```

In words, if there's no input path (1), pick up any records having an empty path. Otherwise (2) if the input path is relative (doesn't start with `/`), make it absolute (prepend `/`) and select any records whose path starts with it. Finally (3) if the path is absolute, collect any records whose path starts with it.

Sure, I can be wrong. I'm just trying to deduce a spec (which we never had!) from the code. IMHO, we should first clarify what the heck a service path is. It looks to me it's just an encoding of a path on a tree? Perhaps a tree used to organise devices in a hierarchy? A formal definition would help here. Once we have that, we can give a precise meaning to "matching a service path"...

Reply:
A service path is a hierarchy: https://fiware-orion.readthedocs.io/en/master/user/service_path/index.html

In the previous implementation, `/a` would select all child of `/a` path, i.e. `/a/b` and `/a/b/c`. The previous query does not work correctly (and return no result) if you ask for `/`, which assuming we want to be consistent, is incorrect.

I would be in favour of introducing `/#` as in context broker, but that would be quite deep change and need to be announced, since people may now rely on `/a` to return subpaths. So goal now it is just to be consistent. if `/a` returns `/a` and `/a/b`, `/` should return `/a`, `/b` and `/a/b`

Reply:
'(/.*)' is the correct filter for the second case

Reply:
yep, it looks right to me too, but I haven't tested it

Reply:
> > https://fiware-orion.readthedocs.io/en/master/user/service_path/index.html
> 
> Ouch. Not the kind of spec I was hoping for. After reading it I've got even more questions than before. There's no clear definition of those terms and it actually looks like that to them, depending on the context, a hierarchy could be a tree or a directed graph (an entity could be under two paths?) but was probably meant to be a poset. Me head hurts :-) Let's forget I ever asked the question and try to fix this based on what QL used to do :-)

there may be two entities with the same id under two different paths, but this are logically two different entities. i think it's explained somewhere. so still a tree, not a graph.
 
> Okay, so `/` shouldn't match the empty path? Or am I missing something?

it should not (also, it should not be possible to have empty path, according to context broker specs, minimal path is always `/`)

Reply:
> here may be two entities with the same id under two different paths, but this are logically two different entities. i think it's explained somewhere. so still a tree, not a graph.

Well, that's one **interpretation** :-) As a matter of fact there's no precise definition and quite a number of statements are seemingly contradictory, e.g. *"It is possible to have an entity with the same ID and type in different Scopes"*, *"Entities belongs to one (and only one) scope."*. Have you seen any statements along the lines of e.g. "a hierarchy is a tree whose leaf nodes are devices...",  "a service path is a string encoding of a root to leaf path whereby..." I don't think we can rely on that spec. To me we should just treat the service path as a sort of entity label which we handle exactly how QL has been doing so far + your root path fix?

> it should not (also, it should not be possible to have empty path, according to context broker specs, minimal path is always /)

What should we do here? We could reject empty service paths when inserting entities but that would be one hell of a breaking change...